### PR TITLE
New SIF Interaction Service

### PR DESCRIPTION
**Pull request by q4wen**
Refs: #916 

Old PC2 graph service is replaced with new SIF graph service. The request returns Json network to the client in 2s (tested with TP53).

### CODE REVIEW

File: src/server/graph-generation/interaction/index.js [hunk 1]
```diff
@@ -19,23 +23,24 @@ function edgeType(type) {
...
     source : geneIds,
-    pattern : ['controls-phosphorylation-of','in-complex-with','controls-expression-of', 'interacts-with'],
-    kind : geneIds.length > 1 ? 'pathsbetween' : 'neighborhood',
-    format : 'txt'
+    pattern : [ 'CONTROLS_PHOSPHORYLATION_OF','IN_COMPLEX_WITH','CONTROLS_EXPRESSION_OF', 'INTERACTS_WITH' ]
   };
 
-  //Fetch graph from PC
-  return pc.query(params).then(res => {
+  const url = ( geneIds.length > 1 ? pathsbetweenUrl : neighborhoodUrl ) + qs.stringify( params );
```
Comment (d2fong) [inline src/server/graph-generation/interaction/index.js]: - you should be using pc.query

File: src/server/graph-generation/interaction/index.js [hunk 2]
```diff
@@ -19,23 +23,24 @@ function edgeType(type) {
...
+  return fetch(url)
```
Comment (d2fong) [inline src/server/graph-generation/interaction/index.js]: All you should be doing is changing the params object to pc.query.

File: src/server/graph-generation/interaction/index.js [hunk 3]
```diff
@@ -51,67 +56,76 @@ const getInteractionGraphFromPC = cache(rawGetInteractionGraphFromPC, pcCache);
...
   };
-  let nodeMap=new Map(); //keeps track of nodes that have already been added
-  if(data){
-    const dataSplit=data.split('\n\n');
-    const nodeMetadata= new Map(dataSplit[1].split('\n').slice(1).map(line =>line.split('\t')).map(line => [line[0], line.slice(1) ]));
-    dataSplit[0].split('\n').slice(1).forEach(line => {
+  let nodeMap = new Map(); //keeps track of nodes that have already been added
+  if( data ){
+
+    const nodeMetadata = null; //TODO: new service does not have node metadata
```
Comment (d2fong) [inline src/server/graph-generation/interaction/index.js]: @IgorRodchenkov  Can you coordinate with Rex on the missing metadata issue?

File: src/server/graph-generation/interaction/index.js (hunk 2 above)
Comment (IgorRodchenkov) [reply src/server/graph-generation/interaction/index.js]: unless you've refactored much there, pc.query uses PC_URL and has nothing to do with the new service (sifgraph); we in fact want to use both PC2 (/get, /traverse, /graph) and Sifgraph endpoints in the app.

**Comment by IgorRodchenkov** [comment]
@d2fong, @q4wen @jvwong 
Wait for me please... I will look into this on Monday.  
If we switch to the sifgraph service, let's have a new method in the pc module rather than messing with pc.query, which is still useful, good to have (for get, traverse queries).

File: src/server/graph-generation/interaction/index.js (hunk 2 above)
Comment (maxkfranz) [reply src/server/graph-generation/interaction/index.js]: The pc module can accept any command and query options, right?  They're just parameters.  The concern with using a separate fetch, as this PR does, is that it's duplicated code and it doesn't accept the PC URL env var.

I think we should be using the existing pc module with a command like 'sifgraph/v1/neighborhood'.

**Comment by IgorRodchenkov** [comment]
No, wrong. Look at the code.

IR

**Comment by IgorRodchenkov** [comment]
My point is that PC is not a single server or endpoint (it could be, but not necessary); pc related services can have different api and urls, so why you are trying to hack to squeeze everything into one pc.query method? We don’t do the same for other services, right?

IR

File: src/server/graph-generation/interaction/index.js [hunk 4]
```diff
@@ -51,67 +56,76 @@ const getInteractionGraphFromPC = cache(rawGetInteractionGraphFromPC, pcCache);
...
+    data.trim().split('\n').forEach( line => {
       const splitLine=line.split('\t');
-      const edgeMetadata = interactionMetadata(splitLine[6],splitLine[4]);
-      addInteraction([splitLine[0], splitLine[2]], splitLine[1], edgeMetadata, network, nodeMap, nodeMetadata, queryIds);
+      const nodes = [splitLine[0], splitLine[2]];
+      const edge = splitLine[1];
+      const mediatorIds = splitLine[4];
```
Comment (jvwong) [inline src/server/graph-generation/interaction/index.js]: I believe this should be `const mediatorIds = splitLine[6];`

File: src/server/graph-generation/interaction/index.js [hunk 5]
```diff
@@ -51,67 +56,76 @@ const getInteractionGraphFromPC = cache(rawGetInteractionGraphFromPC, pcCache);
...
+
+      const edgeMetadata = interactionMetadata( mediatorIds );
```
Comment (jvwong) [inline src/server/graph-generation/interaction/index.js]: You're missing the second parameter, which should be `splitLine[4]` (see above).

**Comment by maxkfranz** [comment]
> My point is that PC is not a single server or endpoint (it could be, but not necessary); pc related services can have different api and urls, so why you are trying to hack to squeeze everything into one pc.query method?

The point I was trying to get across was that the code for the queries should be re-usable.  If your view of the existing pc module is not compatible with the new sif queries, then perhaps different re-usable code is needed.

> We don’t do the same for other services, right?

If a set of services from a single provider all have similar structure, then I would argue that the commonality would naturally lead to a single module for that provider.  You are more familiar with the pc services than I am, so you would be a better judge.  I see two possibilities:

(1) Add to the existing pc module, perhaps adding a parameter for the "service group", e.g. "sifgraph/v1".

(2) Add a new pc-sif module.

I don't really have a preference either way.  My main concerns are that the code is re-usable and that the base URL is configurable, like `PC_URL` is.

**Comment by IgorRodchenkov** [comment]
Why "a new pc-sif module"? It's fine to add new functions to existing server/pathway-commons module, I think.

File: src/server/graph-generation/interaction/index.js (hunk 2 above)
Comment (IgorRodchenkov) [reply src/server/graph-generation/interaction/index.js]: Yeah, I somewhat agree with Dylan and Max; but in order to make it work with like commands cmd:'sifgraph/v1/neighborhood', we have to change the meaning and default value of PC_URL - exclude "pc2/" from it - and do some refactoring (e.g., find/replace 'get' with 'pc2/get', 'search' with 'pc2/search', etc.) through app-ui code.

**Pull request closed without merging**